Add canonical service tags to all trace spans

douglas-reid opened the pull request:
The purpose of this PR is to enable trace query/retrieval by canonical service, without requiring `EnvoyFilter` customization, etc. Canonical services are identified by `mesh_id + namespace + name` (with `revision` providing subsets). 

This PR adds functionality to istiod to tag all trace spans generated by Envoy with the following tags:
- `istio.canonical_service`
- `istio.canonical_revision`
- `istio.namespace`
- `istio.mesh_id`

The values are taken from the environment variables added for this purpose in #23584. Looking forward, this PR could represent the first step in eliminating the need for the environment variables.

This feature can be disabled by setting an env var on pilot-discovery: `PILOT_ENABLE_ISTIO_TAGS=false`. This env var is only meant to be supported during the transition period to a Telemetry API.

[ X ] Policies and Telemetry

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | ef0095fa4c81dd68712965885f2f5c1ce7405dba | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28801/lint_istio/1326614361476698112) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -1734,15 +1734,36 @@ func buildTracingConfig(config *meshconfig.ProxyConfig) *hcm.HttpConnectionManag
 
+func buildCanonicalServiceTags() []*tracing.CustomTag {
+	tags := make([]*tracing.CustomTag, 0, 2)
+	tags = append(tags, &tracing.CustomTag{
```
consider literal
```go
return []*tracing.CustomTag{
                {
                Tag: "canonical_revision",
		Type: &tracing.CustomTag_Environment_{
			Environment: &tracing.CustomTag_Environment{
				Name:         "CANONICAL_REVISION",
				DefaultValue: "latest",
			},
		},
                {
		Tag: "canonical_service",
		Type: &tracing.CustomTag_Environment_{
			Environment: &tracing.CustomTag_Environment{
				Name:         "CANONICAL_SERVICE",
				DefaultValue: "unknown",
			},
		},
	}
}
```

mandarjog replied:
If you define this slice you can use it eveywhere including tests without retyping it.

douglas-reid replied:
Used literal here.

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | ef0095fa4c81dd68712965885f2f5c1ce7405dba | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28801/lint_istio/1326614361476698112) | `/test lint_istio`
integ-operator-controller-tests_istio | ef0095fa4c81dd68712965885f2f5c1ce7405dba | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28801/integ-operator-controller-tests_istio/1326614357890568192) | `/test integ-operator-controller-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -1734,15 +1734,36 @@ func buildTracingConfig(config *meshconfig.ProxyConfig) *hcm.HttpConnectionManag
+		Tag: "canonical_service",
+		Type: &tracing.CustomTag_Environment_{
+			Environment: &tracing.CustomTag_Environment{
+				Name:         "CANONICAL_SERVICE",
```
Where / who substitutes the real value?

douglas-reid replied:
What do you mean by real value?  This tells Envoy to look up `CANONICAL_SERVICE` env and use it. The injector makes sure the proper value is set.

mandarjog reviewed (commented):
Do you want to add additional verification to existing "custom tags" test for cannonical service?
We are considering this foundational.

We also want to add "k8s_cluster" tag

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | ef0095fa4c81dd68712965885f2f5c1ce7405dba | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28801/lint_istio/1326614361476698112) | `/test lint_istio`
integ-operator-controller-tests_istio | ef0095fa4c81dd68712965885f2f5c1ce7405dba | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28801/integ-operator-controller-tests_istio/1326614357890568192) | `/test integ-operator-controller-tests_istio`
integ-telemetry-k8s-tests_istio | ef0095fa4c81dd68712965885f2f5c1ce7405dba | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28801/integ-telemetry-k8s-tests_istio/1326614357643104256) | `/test integ-telemetry-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented:
This is what we have today, k8s_cluster is missing, perhaps other things too.
```yaml
user_agent:	Go-http-client/1.1
http.method:	GET
node_id:	sidecar~10.24.7.127~svc06-0v1-6c68d9b59c-545nd.service-graph06~service-graph06.svc.cluster.local
upstream_cluster:	outbound|8080||svc06-0-3.service-graph06.svc.cluster.local
request_size:	128
zone:	us-central1-a
downstream_cluster:	-
OperationName:	Egress
http.url:	http://svc06-0-3:8080/
http.protocol:	HTTP/1.1
peer.address:	10.24.7.127
response_size:	1024
g.co/agent:	opencensus-cpp [0.5.0-dev]
component:	proxy
http.status_code:	200
guid:x-request-id:	4f358ef5-0e8c-9255-a838-46657230b8ac
```

douglas-reid commented:
> Do you want to add additional verification to existing "custom tags" test for cannonical service?
> We are considering this foundational.

Do you mean, validate that it only appears once?

> We also want to add "k8s_cluster" tag

@gargnupur was going to take a look at this issue. we could add something based on `CLUSTER_ID`, but the question was if we want to capture source & destination cluster, or just the cluster of the current workload.

nrjpoddar reviewed (commented):
How can a user turn this off if they don’t plan to use this tag or don’t want increase their tracing storage?

douglas-reid commented:
> How can a user turn this off if they don’t plan to use this tag or don’t want increase their tracing storage?

Short answer: they can't. This is sorta like how we added new labels to the metrics for the canonical service information. We already lack a way to drop tags that are auto-populated by Envoy, fwiw.

nrjpoddar commented:
IMO these are the kind of things which have been surprising to users as it has downstream impacts on their existing integrations when they upgrade Istio. Can this be behind a global config flag so users can disable if needed with default on?

douglas-reid commented:
> IMO these are the kind of things which have been surprising to users as it has downstream impacts on their existing integrations when they upgrade Istio. Can this be behind a global config flag so users can disable if needed with default on?
> 
> We need to add this in upgrade notice along with release notes.

I understand your point. I'm not sure where to draw the line. We can always add more flags, but I'm not sure that adding yet another field to MeshConfig is really the best path forward.  We could have an env var for pilot like `DISABLE_ISTIO_TAGS` or similar, I guess, as a blanket option. 

FWIW, I think it is OK for Istio to be opinionated about which tags belong in traces. As @mandarjog proposes here, we want to also add cluster info and, I imagine, will later want config-related info. Envoy, for instance, has decided that `downstream_cluster` and `upstream_cluster` belong in each trace, with no way to toggle off. 

@mandarjog @howardjohn thoughts here?

douglas-reid commented:
@nrjpoddar (and all) PTAL. I added a flag for disablement and moved the new tags to have an `istio` prefix. I've also added two more tags (`namespace`, `mesh_id`) to allow proper identification in multi-mesh scenarios, etc.

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-telemetry-k8s-tests_istio | e15abd790a6f58b5811927fb9453b66a942d808f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28801/integ-telemetry-k8s-tests_istio/1327005369167253504) | `/test integ-telemetry-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog reviewed (approved):
Lgtm

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-telemetry-k8s-tests_istio | 0dc08a9269a06658213e2de02075fb4c6355dd52 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28801/integ-telemetry-k8s-tests_istio/1327032406930100224) | `/test integ-telemetry-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nrjpoddar reviewed (commented):
@louiscryan can you weigh in on this? I think of default set of telemetry as some form of stable API that our users have provisioned in their existing environments. Changing that has ramifications on resource and storage costs.

@douglas-reid was nice to add an ENV variable to give users a path to disable it but I still feel like a mesh config option would be better for something like this. What do you think?

nrjpoddar commented:
@douglas-reid I agree about being opinionated but worry about upgrades in general and this is another silent change during upgrades. I like the ENV var, the last thing we need to decide is whether mesh config option is also needed for better UX.

douglas-reid commented:
to add some perspective, we talking about ~100-150 bytes *additive* per span (~1.5MB per 10000 spans). we aren't removing anything or breaking anything. it should not impact upgrade functionality.

adding a meshconfig option that we will then have to deprecate, document, manage, and transition is a real tax. we can incorporate an option in the Telemetry API almost contemporaneously.  The only sticking point is the desire to cherry-pick this onto 1.8 (where Telemetry API will not exist).

nrjpoddar commented on pilot/pkg/features/pilot.go:
```diff
@@ -38,6 +38,15 @@ var (
+	// in trace spans. This is a temporary flag for controlling the feature that will be replaced by
+	// Telemetry API (or accepted as an always-on feature).
+	EnableIstioTags = env.RegisterBoolVar(
+		"PILOT_ENABLE_ISTIO_TAGS",
```
Nit: `PILOT_ENABLE_ISTIO_TAGS` -> `PILOT_ADD_ISTIO_TRACING_TAGS`

kyessenov replied:
Why is pilot even mentioned? ISTIO_ADD_TRACING_TAGS?

douglas-reid replied:
`PILOT` prefix is convention in this file. We have `PILOT_TRACE_SAMPLING` just above and vars like `PILOT_INBOUND_PROTOCOL_DETECTION_TIMEOUT` and `PILOT_HTTP10`. I'm happy to change, but was going for consistency.

mandarjog commented:
I would not worry about UX for a feature that we want to just keep on going forward as always on.

ENV var should suffice. If users complain we can add meshconfig.

nrjpoddar commented:
LGTM

howardjohn commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -1734,15 +1734,53 @@ func buildTracingConfig(config *meshconfig.ProxyConfig) *hcm.HttpConnectionManag
 	return tracingCfg
 }
 
+func defaultTags() []*tracing.CustomTag {
```
Do we have any doc and/or test asserting we are actually setting these env vars? I am just a bit worried about someone like me accidentally removing them since they are "not used"

howardjohn replied:
Looks like we do have a test, didn't finish the PR :slightly_smiling_face:

istio-testing commented:
In response to a cherrypick label: #28801 failed to apply on top of branch "release-1.7":
```
Applying: Add canonical service tags to all trace spans
Using index info to reconstruct a base tree...
M	pilot/pkg/networking/core/v1alpha3/listener.go
M	pilot/pkg/networking/core/v1alpha3/listener_test.go
Falling back to patching base and 3-way merge...
Auto-merging pilot/pkg/networking/core/v1alpha3/listener_test.go
Auto-merging pilot/pkg/networking/core/v1alpha3/listener.go
Applying: Add release note
Applying: Fix telemetry integration tests
Using index info to reconstruct a base tree...
M	pilot/pkg/networking/core/v1alpha3/listener.go
M	tests/integration/telemetry/stackdriver/vm/main_test.go
A	tests/integration/telemetry/stackdriver/vm/testdata/trace.prototext.tmpl
Falling back to patching base and 3-way merge...
CONFLICT (modify/delete): tests/integration/telemetry/stackdriver/vm/testdata/trace.prototext.tmpl deleted in HEAD and modified in Fix telemetry integration tests. Version Fix telemetry integration tests of tests/integration/telemetry/stackdriver/vm/testdata/trace.prototext.tmpl left in tree.
Auto-merging tests/integration/telemetry/stackdriver/vm/main_test.go
CONFLICT (content): Merge conflict in tests/integration/telemetry/stackdriver/vm/main_test.go
Auto-merging pilot/pkg/networking/core/v1alpha3/listener.go
error: Failed to merge in the changes.
hint: Use 'git am --show-current-patch=diff' to see the failed patch
Patch failed at 0003 Fix telemetry integration tests
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".

```

istio-testing commented:
In response to a cherrypick label: new issue created for failed cherrypick: #29019

istio-testing commented:
In response to a cherrypick label: new pull request created: #29020
